client/tinychain/ml/dnn.py

```diff
@@ -12,50 +12,179 @@ def layer(weights, bias, activation):
+            We use property accessor so we can recall self.weights in stead of Dense(self[0])
+            '''
             return Dense(self[0])
 
         def eval(self, inputs):
+            '''
+            eval function applies Activation Function on the dot product 
+            between the inputs  the corresponding weights_matrix + bias
+            type(inputs) :tc.tensor.Dense
+            
+            lets understand einsum :-
+            
+            einsum('ij,ki->kj',[self.weights,inputs]) 
+            
+            let:-
+            ----
+            ij = (3,5)       3 neurons input layer , 5  neurons Next layer
+            ki = (1000,3)    1000 rows of data , 3 neurons this layer
+                             3,5  * 1000,3 -> 1000,5
+            kj   k = 1000 , j = 5
+            so basically einstein sum here is just simple np.dot(self.inputs,weights)
+            '''
+            
             return activation.forward(einsum("ij,ki->kj", [self.weights, inputs])) + self.bias
 
+
         def gradients(self, A_prev, dA, Z):
+            
+            '''
+            gradients function :it calculates delta values in Weights , Bias and in previous layer.
+            A_prev : Previous Input.
+            Z : Current Output before Activation.
+            dA: detla_Error.
+            dZ : delta Error in the Current Layer.
+            dA_Prev: delta Error in Previous Layer.
+            d_weights: delta Weights.
+            d_bias : delta Bias.
+            activation.backward(dA,Z) : performs derivative version of your activation function
+                                          to apply backward propagation.
+            
+            '''
             dZ = activation.backward(dA, Z).copy()
-            dA_prev = einsum("kj,ij->ki", [dZ, self[0]])
+            dA_prev = einsum("kj,ij->ki", [dZ, self.weights])  
             d_weights = einsum("kj,ki->ij", [dZ, A_prev])
             d_bias = dZ.sum(0)
             return dA_prev, d_weights, d_bias
 
         def train_eval(self, inputs):
+            '''
+            train_eval function you can use it on any layer to get its
+            output after dot product and after applying activation function on that output
+            
+            Z : Output of dot product between weight and inputs .
+            A : returning the output after applying activation function on it.
+            '''
             Z = einsum("ij,ki->kj", [self.weights, inputs])
             A = activation.forward(Z) + self.bias
             return A, Z
 
         def update(self, d_weights, d_bias):
+            '''
+            update function is used to ensure that update operations are done
+            New_Weight = Old Weight - Delta Weight
+            New_Bias = Old Bias - Delta Bias
+            '''
             return self.weights.write(self.weights - d_weights), self.bias.write(self.bias - d_bias)
+                
 
     return DNNLayer([weights, bias])
 
 
+def optimizers():
```

Comment:
This function has no parameters. Why is it necessary?

Reply:
Is it okay to build Adam Class without containing it by function ? 
I thought its part of the way OOP works in tinychain

Reply:
Sometimes a class implementation is easier or more efficient when it's limited by parameters known at encoding time, like the number of layers in `def neural_net(layers)` but in this case, since you don't have any encoding-time parameters, you don't need the function.

Reply:
cool then i will erase it